hw/ip/pwm/doc/_index.md

```diff
@@ -0,0 +1,313 @@
+
+The PWM IP provides a hardware-controlled blink feature, which can periodically toggle the output between two states with separate duty cycles.
+This feature can be used to significantly reduce the output duty cycle (blink-off), or, in the case of tri-color LEDs, toggle the apparent LED color between the two settings.
+This feature is enabled on a per-channel basis, and for each blinking channel, the blink period and blink duty cycle are programmed in terms of the number of pulses generated in either state.
+
+As a variant of the blink feature, the output duty cycle can also be programmed to linearly increase and decrease in time.
+In this "heartbeat" mode, the duty cycle increments by a programmable amount after a programmable number of pulses, starting at some firmware-selected initial duty cycle.
+Once the internal duty cycle reaches the target value, the internal duty cycle begins to decrement until it returns to the initial value, at which point the cycle repeats until heartbeat mode is disabled.
+
+# Theory of Operations
+
+## Block Diagram
+
+![](pwm_block_diagram.svg)
+
+## Hardware Interfaces
+
+{{< hwcfg "hw/ip/pwm/data/pwm.hjson" >}}
+
+## Design Details
+
+### Phase and Duty Cycle Representation
+
+The PWM IP creates series of pulses with the desired on-off duty cycle.
+The duty cycle, DC, is typically expressed a fraction of pulse duration, <i>d</i>, over the period between pulses, <i>T</i>:
+$$DC\equiv d/T$$.
+Since 0&lt;<i>d</i>&lt;<i>T</i>, the duty cycle generally ranges from 0 to 1.
+
+The PWM IP can control the duty cycle in a number of ways:
+- The PWM can be programmed to generate pulses at a firmware-defined duty cycle.
+- The duty cycle can be programmed to toggle (or "blink") between two duty cycles at a programmable rate.
+In the tri-color LED use-case, this would have the visual effect of making the LED blink between two colors.
+- The duty cycle can linearly sweep in time, gradually shifting back-and-forth between two endpoints.
+
+Thus the duty cycle can be controlled by firmware, or may change under hardware control.
+The operation of each of these modes will be discussed later.
+
+Each channel can also be assigned a different phase delay.
+Like the duty cycle, this delay is expressed as a fraction of the pulse period, T.
+The phase delay of each channel is always directly controlled by a firmware register value.
+
+Since the phase and duty cycle are always a fraction less than or equal to one, the PWM IP represents them as 16-bit fixed point numbers, with an implicit 16-bit shift.
+If the duty cycle is internally represented as a 16-bit value x, the output pulse train will have the duty cycle:
+$$DC(x)=\frac{x}{2^{16}}.$$
+Thus the allowed duty cycle in principle ranges from 0 to 1-1/2<sup>16</sup>.
```

Comment:
This would probably be a bit easier to read with parentheses around `1 - 1/2^16` (otherwise it looks a bit like a range of values `a - b`)

Reply:
Ah.  I see.  Sure.